Add kdeploy script

lucasdemarchi opened the pull request:
Helper script to deploy kernel and/or modules to a host.

lucasdemarchi committed:
Add kdeploy script

Helper script to deploy kernel and/or modules to a host.

Signed-off-by: Lucas De Marchi <lucas.de.marchi@gmail.com>

lucasdemarchi commented:
@guludo could you take a look on this? I'm successfully using it to deploy kernel and modules to a VM and this is much faster than my previous approach with `custom-kernel-{gen,install}` in this same repository.

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	. /etc/os-release
+
+	if [[ "$ID_LIKE" == "arch" ]]; then
+		mkinitcpio -k "/boot/vmlinuz-$k -g" "/boot/initramfs-$k.img"
```
The way this command is spelled, `-g` becomes part of the argument to `-k`.

lucasdemarchi replied:
thanks, fixed

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	# shellcheck disable=SC1091
+	. /etc/os-release
+
+	if [[ "$ID_LIKE" == "arch" ]]; then
```
Hm. I'm on Arch and my os-release file does not define ID_LIKE:

```console
$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo
```

lucasdemarchi replied:
thanks, added a check for ID too

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	shift
+	local key val="" n=0
+
+	IFS="=" read -r key val <<< "$1"
```
What about an option without redirects?

```bash
[[ "$1" =~ ^([^=]+)(=(.*)) ]]
key=${BASH_REMATCH[1]}
val=${BASH_REMATCH[3]}
```

lucasdemarchi replied:
seems good... but any advantage on doing that?

guludo replied:
Avoiding the redirect?

I'm not sure if bash is smart enough to realize that the builtin read command can be completed without opening any fd for reading the data.

guludo replied:
It looks like it does create a pipe:

```console
$ strace -o /tmp/strace bash -c 'IFS="=" read -r bla ble <<< "x=y"'
$ grep -C5 'write(.*x=y.*)' /tmp/strace
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
getpeername(0, 0x7ffe0af51f60, [16])    = -1 ENOTSOCK (Socket operation on non-socket)
rt_sigprocmask(SIG_BLOCK, NULL, [], 8)  = 0
pipe2([3, 4], 0)                        = 0
fcntl(4, F_GETPIPE_SZ)                  = 65536
write(4, "x=y\n", 4)                    = 4
close(4)                                = 0
fcntl(0, F_GETFD)                       = 0
fcntl(0, F_DUPFD, 10)                   = 10
fcntl(0, F_GETFD)                       = 0
fcntl(10, F_SETFD, FD_CLOEXEC)          = 0
```

lucasdemarchi replied:
ok... I had to rework this to get it working with `set -u` this script is using. Now apparently it's doing the right thing without/without "="

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+
+	OPTIONS:
+	    -h, --help			Show this help
+	    -i, --include[=] GLOB	Include only specific module glob
```
Looks like --include is not the only options that allows "=".

lucasdemarchi replied:
fixed

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	local key val="" n=0
+
+	IFS="=" read -r key val <<< "$1"
+	if [[ -z "$val" && $# -gt 1 ]]; then
```
Maybe error out complaining that $key requires an argument if $# is zero?

lucasdemarchi replied:
fixed

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+			OPT_REMOTE_PORT="${args[2]}"
+			;;
+		*)
+			argerr "Unknown option $1"
```
Where is argerr?

lucasdemarchi replied:
forgot to paste in the script... but it's the function you wrote for one of our scripts :)

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+
+KREL=$(make kernelrelease)
+
+if [[ -n "$OPT_REMOTE_HOST" && "$DESTDIR" == "/" ]]; then
```
For the remote case, should we really depend on the value of DESTDIR to use the temporary directory?

lucasdemarchi replied:
It was intentional, yes... because I want to be able to do:  `DEST=tmpinstall kdeploy ....` if I want to inspect what is the directory tree, without the script removing the temp directory.

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+
+mkdir -p "$DESTDIR/boot/"
+# TODO: figure out arch
+cp arch/x86/boot/bzImage "$DESTDIR/boot/vmlinuz-$KREL"
```
Shouldn't this honor OPT_MODULE_ONLY and OPT_MODULE_INCLUDE?

Looks like we are checking those variables only for the remote case, but I think it would be good to make them usable for the local case as well.

I think we could:

1. Do the local "install" that honors all flags.
2. If OPT_REMOTE_HOST is present, then rsync from the temporary destdir to the remote host.

lucasdemarchi replied:
humn... I think the problem is the OPT_MODULE_INCLUDE for local case. I think we'd always need the tmp dir for supporting that.

lucasdemarchi replied:
and just noticed it doesn't work locally today anymore if the script is not executed as root. I think I will require the tmpdir to make things simpler and then also support the --include in this case.

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	# shellcheck disable=SC1091
+	. /etc/os-release
+
+	update-grub
```
I would check if grub is installed first.

lucasdemarchi replied:
but then the only sane thing to do is to error out, no?  I thought about supporting multiple options, but I didn't find a reliable way (yet).  I was parsing `efibootmgr -v` but then it occurred to me it wouldn't work if we have a chainload: we have some machines in the lab that do   refind -> grub. We don't want to update refind, we want to update grub.

So... I decided to keep it simple for now and just support grub.

lucasdemarchi replied:
Another option is to delegate to `kernel-install` like I was doing in other scripts... however it doesn't work out of the box for me with grub in both arch and ubuntu:

```
tree /usr/lib/kernel/install.d/
/usr/lib/kernel/install.d/
├── 50-depmod.install
├── 50-mkinitcpio.install
├── 90-loaderentry.install
└── 90-uki-copy.install

1 directory, 4 files
```

90-loaderentry.install  will bail out if $KERNEL_INSTALL_LAYOUT is not bls

guludo replied:
> but then the only sane thing to do is to error out, no?

Or warn that we are not reconfiguring any bootloader.

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	if [[ "$OPT_MODULE_ONLY" == "no" ]]; then
+		typeset -f initrd -f bootloader -f sudo-function | \
+			ssh -p "$OPT_REMOTE_PORT" "$OPT_REMOTE_USER@$OPT_REMOTE_HOST" \
+			"$(cat); sudo-function initrd $KREL && sudo-function bootloader"
```
Hm... I think this works, but wouldn't it be simpler to:

```bash
ssh -p "$OPT_REMOTE_PORT" "$OPT_REMOTE_USER@$OPT_REMOTE_HOST" \
    "sudo bash -c 'source /dev/stdin; initrd $KREL && bootloader'" < "${BASH_SOURCE[0]}"
```

With that:

- we can get rid of the sudo-function;
- we don't need to worry about maintaining the list of functions that need to be exported.

Of course, for that to work, we would also need to ensure that the "main execution" is encapsulated into a main function and gets executed only on the local machine.

guludo replied:
Also, because we are running sudo on the remote, wouldn't the option `-t` be required for the ssh command?

lucasdemarchi replied:
I don't see any advantage on sending the entire script... particularly because it would execute things that it shouldn't, unless we are careful to encapsulate everything. Also the functions should still be self-contained, not relying on global variables. 

I like sending only the functions needed  -  and the reason to annotate them as "can be remotely executed" is the need to be self-contained.

Regarding `-t`, dunno... it worked for me without.

guludo replied:
> Regarding -t, dunno... it worked for me without.

I remember needing a pseudoterminal in some machines. Maybe this is somehow related to how sudo is configured.

guludo replied:
> I don't see any advantage on sending the entire script... particularly because it would execute things that it shouldn't, unless we are careful to encapsulate everything. Also the functions should still be self-contained, not relying on global variables.
> 
> I like sending only the functions needed - and the reason to annotate them as "can be remotely executed" is the need to be self-contained.

Okay. I just don't like much the fact that this solution requires us to remember to make those functions self-contained or to also include any dependencies that might arise in the future. But that works.

lucasdemarchi replied:
> > Regarding -t, dunno... it worked for me without.

> I remember needing a pseudoterminal in some machines. Maybe this is somehow related to how sudo is configured.

Humn... that's because currently I use the script with the remote user in the sudoers without asking the password. This is not very good, but I'm not sure yet how to make it simpler without asking the passwd multiple times and having to copy files to a temporary location in the remote.

lucasdemarchi replied:
I will leave this for future

guludo commented on kdeploy:
```diff
@@ -0,0 +1,186 @@
+	fi
+elif [[ "$DESTDIR" == "/" && "$OPT_MODULE_ONLY" == "no" ]]; then
+	sudo-function initrd "$KREL"
+	sudeo-function bootloader
```
s/sudeo/sudo/

lucasdemarchi replied:
fixed, thanks.

guludo commented on kdeploy:
```diff
@@ -0,0 +1,206 @@
+	local key val n=0
+
+	[[ "$1" =~ ^([^=]+)(=(.*)) ]] || true
+	if [[ -z ${BASH_REMATCH+x} ]]; then
```
Hm... Here you coud use:

```bash
if [[ "$1" =~ ^([^=]+)(=(.*)) ]]; then
```

and switch the "then" and "else" blocks.

guludo reviewed (approved):
I didn't try running this script myself, but it looks good to me.

I guess https://github.com/lucasdemarchi/toolbox/pull/33#discussion_r1783196067 could be addressed as a follow up work.

lucasdemarchi commented:
pushed again with more modularized functions. Also added a few useful things:  `--remote-reboot`, `--wait-remote-reboot` and moved DESTDIR to actually be --tmpdir=..... With this now we always install to a tmpdir and sync with the remote/local host. I tested this only with remote host, need to test locally.

Developing with qemu on my local machine and submitting a kernel built on $PWD:

```
$ time kdeploy -u test -r localhost -p 4022 --wait-remote-reboot
...
real    0m21.304s
user    0m2.846s
```

Wait logic can probably be improved as I see it ready a few seconds before the second connect attempt.
